Transition

PoiSenpai opened the pull request:
Added Transition mechanic
Added void gate entity
Added limitcut marker condition
Updated CircleBladeMelusine, NerveGasKaliya, and RepellingCannonADS to contain delays

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Attacks/VoidGate.cs:
```diff
@@ -0,0 +1,124 @@
+    public static Entity CreateEntity(World world)
+    {
+        return world.Entity()
+            .Set(new Model(-1))
+            .Set(new Position())
+            .Set(new Rotation())
+            .Set(new Scale())
+            .Set(new UniformScale(1f))
```
Use FakeActor.Create()

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Encounters/UCOB/Transition.cs:
```diff
@@ -0,0 +1,309 @@
+                DebugOutput();
+                break;
+            case Phase.Teraflare:
+                localPlayer = this.Dalamud.ClientState.LocalPlayer;
```
Player shuffling should happen here

PoiSenpai commented on DalamudPlugin/RaidsRewritten/Scripts/Encounters/UCOB/Transition.cs:
```diff
@@ -0,0 +1,309 @@
+
+                var da = DelayedAction.Create(World, () => 
+                { 
+                    ShowAds(telegraphs[0], TelegraphDelay); 
```
This telegraph is randomized as we are randomizing the telegraph array to prevent parity between numbers/telegraphs. the 0th entity was just picked because the array is already randomized

PoiSenpai commented:
Looking into an issue that is having the logic desync on the replay.

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Encounters/UCOB/Transition.cs:
```diff
@@ -242,15 +258,21 @@ public override void OnActionEffectEvent(ActionEffectSet set)
-                        LimitCutNumber.ApplyToTarget(fa, 10, SymbolPaths[playerNumber]);
+                        var lc = this.World.Entity()
+                            .Set(new ActorVfx(SymbolPaths[SymbolNumber[playerNumber]]))
+                            .ChildOf(fa);
```
You don't need to make a fake actor under the player, put the ActorVfx entity under the Player entity

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Encounters/UCOB/Transition.cs:
```diff
@@ -242,15 +258,21 @@ public override void OnActionEffectEvent(ActionEffectSet set)
+                            .Set(new ActorVfx(SymbolPaths[SymbolNumber[resolution]]))
+                            .ChildOf(e);
+                        attacks.Add(lc);
                     });
```
Try to avoid using .ForEach when you can write a foreach loop instead

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Encounters/UCOB/Transition.cs:
```diff
@@ -197,22 +210,25 @@ public override void OnActionEffectEvent(ActionEffectSet set)
+                Shuffle(random, playerList);
                 int playerNumber = playerList.IndexOf(localPlayer);
-
+                Logger.Debug($"{playerNumber}");
```
Cleanup the logging when you're ready to merge

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Encounters/UCOB/UcobRewritten.cs:
```diff
@@ -310,10 +310,10 @@ public void DrawConfig()
 
-        bool cobTransition = configuration.GetEncounterSetting(CobTransitionKey, this.defaultBoolSettings[CobTransitionKey]);
+        bool cobTransition = configuration.GetEncounterSetting(JunctionCoilsKey, this.defaultBoolSettings[JunctionCoilsKey]);
         if (ImGui.Checkbox("Transition", ref cobTransition))
```
```suggestion
        if (ImGui.Checkbox("Junction Coils", ref cobTransition))
```

PoiSenpai commented:
Penumbra module is preventing pull on local branch. 
Updated local build to represent the Junction Coil rename for UI. 
Updated variable names in UcobRewritten to align with term update
Removed random check OnActionEffectEvent in JunctionCoils and added initializations to random in each case.

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Attacks/VoidGate.cs:
```diff
@@ -0,0 +1,123 @@
+                    case Phase.Spawn:
+                        if (component.ElapsedTime < spawnDelay.DelayTime) { return; }
+                        AddActorVfx(animationActor, AbsorbVfx);
+                        DelayedAction.Create(world, () =>
+                        {
+                            AddActorVfx(gateActor, GateActorVfx);
+                        }, 0.5f);
```
Delayed actions that operate on a system's entity need to be a child of the entity, or else if the entity is deleted before the delayed action executes, ECS will throw an exception crash
```suggestion
                        DelayedAction.Create(world, () =>
                        {
                            AddActorVfx(gateActor, GateActorVfx);
                        }, 0.5f).ChildOf(gateActor);
```

Ricimon commented on DalamudPlugin/RaidsRewritten/Scripts/Attacks/VoidGate.cs:
```diff
@@ -0,0 +1,123 @@
+
+                        AddActorVfx(animationActor, ExpelVfx);
+                        
+                        DelayedAction.Create(world, () =>
+                        {
+                            gateActor.Destruct();
+                        }, 0.5f);
```
```suggestion
                        DelayedAction.Create(world, () =>
                        {
                            gateActor.Destruct();
                        }, 0.5f).ChildOf(gateActor);
```
